src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs

```diff
@@ -504,17 +500,38 @@ private static void RemoveDirectoryInternal(DirectoryInfo directory, bool recurs
                     case Interop.Error.EPERM:
                     case Interop.Error.EROFS:
                     case Interop.Error.EISDIR:
-                        throw new IOException(SR.Format(SR.UnauthorizedAccess_IODenied_Path, directory.FullName)); // match Win32 exception
+                        throw new IOException(SR.Format(SR.UnauthorizedAccess_IODenied_Path, fullPath)); // match Win32 exception
                     case Interop.Error.ENOENT:
-                        if (!throwOnTopLevelDirectoryNotFound)
+                        if (!topLevel)
                         {
-                            return;
+                            return true;
+                        }
+                        goto default;
+                    case Interop.Error.ENOTDIR:
+                        if (topLevel)
+                        {
+                            if (!DirectoryExists(fullPath))
```

Comment:
I don't understand why `DirectoryExists` is called for `ENOTDIR`. According to the [manual](https://man7.org/linux/man-pages/man2/rmdir.2.html)

rmdir fails with:
> * `ENOENT` A directory component in pathname does not exist or is a dangling symbolic link
> * `ENOTDIR` - pathname, or a component used as a directory in pathname, is not, in fact, a directory

I would expect that:
* on `ENOENT` we check if it's dangling symbolic link. If it is, we remove the link.
* on `ENOTDIR` we just call `DeleteFile`

Reply:
See https://github.com/dotnet/runtime/pull/59520#discussion_r752447729.

Reply:
I've added a comment.